presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java

```diff
@@ -0,0 +1,110 @@
+import com.facebook.presto.spi.type.Type;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.spi.type.StandardTypes.ARRAY;
+import static com.facebook.presto.spi.type.StandardTypes.MAP;
+import static com.facebook.presto.spi.type.StandardTypes.ROW;
+import static com.facebook.presto.spi.type.VarbinaryType.VARBINARY;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static io.airlift.slice.Slices.utf8Slice;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+
+public final class ElasticsearchUtils
+{
+    private ElasticsearchUtils() {}
+
+    public static Block serializeObject(Type type, BlockBuilder builder, Object object)
+    {
+        if (ROW.equals(type.getTypeSignature().getBase())) {
+            return serializeStruct(type, builder, object);
+        }
+        if (MAP.equals(type.getTypeSignature().getBase()) || ARRAY.equals(type.getTypeSignature().getBase())) {
+            throw new IllegalArgumentException("Type not supported: " + type.getDisplayName());
+        }
+        return serializePrimitive(type, builder, object);
+    }
+
+    private static Block serializeStruct(Type type, BlockBuilder builder, Object object)
+    {
+        if (object == null) {
+            requireNonNull(builder, "builder is null");
+            builder.appendNull();
+            return builder.build();
+        }
+
+        if (builder == null) {
+            builder = type.createBlockBuilder(null, 1);
+        }
+
+        BlockBuilder currentBuilder = builder.beginBlockEntry();
+        List<Type> typeParameters = type.getTypeParameters();
+
+        for (int i = 0; i < typeParameters.size(); i++) {
+            Optional<String> fieldName = type.getTypeSignature().getParameters().get(i).getNamedTypeSignature().getName();
+            String name = "";
+            if (fieldName.isPresent()) {
+                name = fieldName.get().toLowerCase(ENGLISH);
+            }
+            Object value = ((Map) object).get(name);
+            serializeObject(typeParameters.get(i), currentBuilder, value);
+        }
+
+        builder.closeEntry();
+        return (Block) type.getObject(builder, 0);
+    }
+
+    private static Block serializePrimitive(Type type, BlockBuilder builder, Object object)
+    {
+        requireNonNull(builder, "builder is null");
+
+        if (object == null) {
+            builder.appendNull();
+            return builder.build();
+        }
+
+        if (type.equals(BOOLEAN)) {
+            type.writeBoolean(builder, (Boolean) object);
+        }
+        else if (type.equals(BIGINT)) {
+            type.writeLong(builder, (Long) object);
```

Comment:
At times, we see the data type of object as INTEGER. I don't know whether ES is storing the data as int to save space when the value is small. 
`                if (object instanceof Integer) {
                    type.writeLong(builder, Long.valueOf((Integer) object).longValue());
                }
                else {
                    type.writeLong(builder, (Long) object);
                }`
I added this to handle that situation.

Reply:
@zhenxiao @sajjoseph if this change is needed can you please open a separate PR?

Reply:
I will.
There are few other issues (including a race condition while using the retry mechanism) I came across and I am working on a fix. I will create a PR with those changes.